### DISCUSSION TITLE
bring back classic skinfolders for lazer or create a symlink possibility

### DISCUSSION THREAD

**Discussion by Fletscherrrr** [ppy/osu#24241] [Ideas] [answered]
So if I remember correctly hashing most of the game related files was done in order to not have users fiddling around where they shouldn't. However especially in the case of skinning I think this should be reconsidered. There is one major reason why I think so.

Its simply inconvenient. Here is why I think so:
Creating a skin involves a lot of testing, refreshing and changing elements in its folder. Even if the majority of the doesn't create their own skin most people still have a handful of changes/personalizations they want to include in a skin and this is where I prefer stable. You can simply open the skin folder, drag files into it and go on with your day.
In case of skinning elements, lets say something went wrong during the process or you just want to push a hotfix to an element.
You export the new element directly into the desired skin folder, reload and its done.

How does this process work in Lazer? I have no idea. 
- .osk files vanish after they have been loaded into the game
- you cant directly access skin related files because everything is hashed 

So is there even a convenient way that makes gives the ability to push a hotfix to elements that are wip other than deleting a skin and create a .osk file over and over again until they're good? 

If yes I'd love to hear about it because at the moment this way of handling files is one of the biggest dealbreakers for me.

**Comment by bdach** [comment]
The skin editor should eventually support sprite substitution. As a last case scenario we can consider adding something that's already planned for beatmap editor (https://github.com/ppy/osu/issues/18902) but for skins.

**Reply by Fletscherrrr** [reply]
Honestly sprite substitution doesn't really seem like it would make it much more convenient.
Especially when you need to change a lot of stuff as mentioned before. I guess I just don't see it yet.
Just directly pasting something into a folder however is just the workflow everyone is used to so I don't to immediately say its worse than stable just because I don't know it yet.
Though I would say sometimes things shouldn't change that drastically. Gotta see what it's gonna be like.

**Reply by peppy** [reply, marked as answer]
FWIW you can already drag anything into the osu! window while in the skin editor. it will update your skin. Just delete the sprites that appear on screen afterwards.

**Reply by Fletscherrrr** [reply]
welp gotta wait and see about the functionality when its actually starting to become a thing and skinning generally gets more focus

**Reply by peppy** [reply]
For anyone confused on the above process, here's an example:

https://github.com/ppy/osu/assets/191335/7a39047c-d21b-4b42-a82f-e4cba6ed2e4d

**Reply by dravpin** [reply]
This doesnt work ive tried every combination of operation order and file name all of which result in a png on the screen and no replacement cursor where do i find skinning resources? as it seems like this whole editor is undocumented and this is the only thing i can find relating to lazer skinning

**Reply by Joehuu** [reply]
You currently need to change to another skin and back.

Another way is to edit the files directly by `File` -> `Edit externally`.

**Reply by dravpin** [reply]
> You currently need to change to another skin and back.
> 
> Another way is to edit the files directly by `File` -> `Edit externally`.

doesnt work either i tried edit externally placing a png named cursor@2x.png then clicking finish editing then switiching in and out of the skin as well as restarting the game non of which worked

**Reply by Joehuu** [reply]
Note that you can't edit the cursor on "argon" or "triangles" skins right now. You can only edit it on "classic" / "retro" / any other osu!(stable) skins.

**Reply by dravpin** [reply]
> Note that you can't edit the cursor on "argon" or "triangles" skins right now. You can only edit it on "classic" / "retro" / any other osu!(stable) skins.

Thats a little random of a limitation i like the argon skin and would like to use it as a base but just cant because? as well as how do i know what i can and cant modify as again i cant find any resources to skinning lazer

**Reply by Aworldc** [reply]
> > Note that you can't edit the cursor on "argon" or "triangles" skins right now. You can only edit it on "classic" / "retro" / any other osu!(stable) skins.
> 
> Thats a little random of a limitation i like the argon skin and would like to use it as a base but just cant because? as well as how do i know what i can and cant modify as again i cant find any resources to skinning lazer

Yeah ittd be nice if argon was just another stable skin you could export and modify.

**Comment by ItsCollector** [comment]
i agree, currently it's very inconvenient to do anything skinning related as I have to constantly "edit externally" for small tweaks and that requires multiple clicks through various menus, including leaving the editor because the skin preview doesn't automatically refresh with the change for some reason + the preview is not accurate for modes like osu!mania. 

it's easier to edit skins on stable then import them into lazer when I'm happy with the changes.

I can't comment on other modes but for mania
- The answer described by peppy with mania doesn't work for 95%+ of osu!mania skins because the files are often in sub-folders with different names, and the skin.ini points to those files, so you cannot change those without opening externally. 
- You still can't edit the skin.ini within the editor without editing externally.
